### ISSUE TITLE
different results in parallel + error in output

### ISSUE THREAD

**Issue by francispoulin** [spectralDNS/shenfun#49]
Maybe I should have split this up but both happen when I go from one course to two, so maybe they are connected.

First, the results in serial look great and are way faster than was I was doing previously. I think it goes to show that RK4 is much faster than Adams Bashforth 3.  

Unfortunately, when I run it in parallel i get two odd results.

First, the writing seems to freeze and I need to kill it.  The results are copied below.

Second, I am looking at a double jet and their initial conditions are almost identical except for a random perturbation, but as one can see from a snapshot they evolve differently.  I would say one is more damped then the other, which might suggest something to do with the padding.  But I tried it without padding and it did not resolve the issue.  Unfortunately, it's not clear to me how to include a figure in this post but I am happy to do that would help.

Clearly I've messed something up but not sure what.  Any suggestions what I should look into?

`# error

/home/fpoulin/software/anaconda3/envs/shenfun/lib/python3.8/site-packages/mpi4py_fft/io/generate_xdmf.py:117: H5pyDeprecationWarning: The default file mode will change to 'r' (read-only) in h5py 3.0. To suppress this warning, pass the mode you need to h5py.File(), or set the global default h5.get_config().default_file_mode, or set the environment variable H5PY_DEFAULT_READONLY=1. Available modes are: 'r', 'r+', 'w', 'w-'/'x', 'a'. See the docs for details.
  f = h5py.File(h5filename)
Traceback (most recent call last):
  File "/home/fpoulin/software/anaconda3/envs/shenfun/lib/python3.8/site-packages/h5py/_hl/files.py", line 199, in make_fid
    fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 88, in h5py.h5f.open
OSError: Unable to open file (unable to lock file, errno = 11, error message = 'Resource temporarily unavailable')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/fpoulin/software/anaconda3/envs/shenfun/lib/python3.8/site-packages/h5py/_hl/files.py", line 202, in make_fid
    fid = h5f.open(name, h5f.ACC_RDONLY, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 88, in h5py.h5f.open
OSError: Unable to open file (unable to lock file, errno = 11, error message = 'Resource temporarily unavailable')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "NavierStokes2D.py", line 141, in <module>
    fftw.export_wisdom('QG1L.wisdom')
  File "/home/fpoulin/software/anaconda3/envs/shenfun/lib/python3.8/site-packages/mpi4py_fft/io/generate_xdmf.py", line 117, in generate_xdmf
    f = h5py.File(h5filename)
  File "/home/fpoulin/software/anaconda3/envs/shenfun/lib/python3.8/site-packages/h5py/_hl/files.py", line 406, in __init__
    fid = make_fid(name, mode, userblock_size,
  File "/home/fpoulin/software/anaconda3/envs/shenfun/lib/python3.8/site-packages/h5py/_hl/files.py", line 204, in make_fid
    fid = h5f.create(name, h5f.ACC_EXCL, fapl=fapl, fcpl=fcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 108, in h5py.h5f.create
OSError: Unable to create file (unable to open file: name = 'QG_1Layer_256.h5', errno = 17, error message = 'File exists', flags = 15, o_flags = c2)
`

**Comment by mikaem** [comment]
Hi Francis,

I need some more input to be able to help. If you could create a minimum reproducing script and attach it or link to it, then I could have a look.

**Comment by francispoulin** [comment]
Hello Mikael,

Thanks.  I am including two images for a very coarse run of 128x128.

The first is the result of a serial run, where we get three modes in the channel direction.  

The second is what arises when I run it on two cores.  It seems that the parallelization is in the periodic direction.  On the right we have fewer vortices because it seems to have cut the domain in half.  The left has more vortices, and they are clearly more stable as we have them much tighter.  

Since this is solving the 2D vorticity equation, which is pretty basic, I will copy this code below, hoping that this is simple enough.

Thanks in advance for yoru help and sorry that I have not been able to figure it out.

![Screenshot from 2020-05-11 15-26-43](https://user-images.githubusercontent.com/8239041/81603320-226d6f80-939c-11ea-90be-034db8cc93dc.png)

![Screenshot from 2020-05-11 15-26-03](https://user-images.githubusercontent.com/8239041/81603296-18e40780-939c-11ea-9bb5-2b997ac2b28f.png)

**Comment by francispoulin** [comment]
`from mpi4py import MPI
    import numpy as np
    from shenfun import *
    from mpi4py_fft import generate_xdmf, fftw

    import matplotlib.pyplot as plt
    import time
    import sys

    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()

    # Plot on the fly
    display = True #False

    # Time parameters
    t_domain = (0.0, 100.0)
    dt       = 1e-2

    # Geometry
    M = 7
    N = (2**M, 2**M)
    L = [20.0, 40.]

    # Funcion Spaces
    V1  = Basis(N[0], 'F', dtype='D', domain=(0, L[0]))
    V2  = Basis(N[1], 'F', dtype='d', domain=(0, L[1]))
    T   = TensorProductSpace(comm, (V1, V2), **{'planner_effort': 'FFTW_MEASURE'})
    TV  = VectorTensorProductSpace(T)

    # Padded Spaces
    Tp  = T.get_dealiased((1.5, 1.5))
    TVp = TV.get_dealiased((1.5, 1.5))

    # Try to import wisdom. 
    try:
        fftw.import_wisdom('QG1L.wisdom')
        print('Importing wisdom')
    except:
        print('No wisdom imported')

    q_hat,     q     = Function(T),  Array(T)           # potential vorticity (PV)
    gradq_hat, gradq = Function(TV), Array(TV)          # gradient of PV
    qb  = Array(T)
    qp, gradqp, Up = Array(Tp), Array(TVp), Array(TVp)  # Padded arrays

    # Mesh
    X  = T.local_mesh(True)

    # Wavenumbers
    K  = np.array(T.local_wavenumbers(True,True))
    K2 = np.sum(K*K, 0, dtype=float)
    K_over_K2 = K.astype(float) / np.where(K2 == 0, 1, K2).astype(float)

    # Initialization
    np.random.seed(1)
    #np.random.seed(rank)
    q_pert = 1e-4*2.*(np.random.random(q.shape).astype(q.dtype)-0.5)
    qb[:] = -2.*np.tanh(X[1] - L[1]/4.)/pow(np.cosh(X[1] - L[1]/4.),2) \
            + 2.*np.tanh(X[1] - 3*L[1]/4.)/pow(np.cosh(X[1] - 3.*L[1]/4.),2)
    q[:] = qb + q_pert
    q_hat = T.forward(q, q_hat)

    def Flux_QG(self, f, f_hat, rhs_hat, **par):
    
        global T, Tp, TVp, Up, gradqp, gradq_hat#, qb, Ub, 
    
        # Gradient of q
        gradq_hat = 1j*K*f_hat
        gradqp    = TVp.backward(gradq_hat, gradqp)

        # Velocity
        Up[0]  = Tp.backward( 1j*K_over_K2[1]*q_hat, Up[0])
        Up[1]  = Tp.backward(-1j*K_over_K2[0]*q_hat, Up[1])

        # Flux
        rhs  = -(Up[0]*gradqp[0] + Up[1]*gradqp[1])
        rhs_hat = Tp.forward(rhs, rhs_hat)

        return rhs_hat    

    if display:
        # Create figure
        plt.figure(figsize=(10,10))
        image = plt.contourf(X[0], X[1], q - 0*qb, 100)
        image.ax.set_title('q at t = %6.2f' % t_domain[0])
        #plt.colorbar()
        plt.draw()
        plt.pause(1e-6)

    count = 0

    def update(self, q, q_hat, t, tstep, plot_tstep, write_tstep, file, **par):
        global count
        if tstep % plot_tstep == 0 and plot_tstep > 0:
            if display:
                q = T.backward(q_hat, q)
                image.ax.clear()
                image.ax.contourf(X[0], X[1], q - 0*qb, 100)
                image.ax.set_title('q at t = %6.2f' % t)
                #plt.colorbar()
                plt.pause(1e-6)
            count += 1
            #plt.savefig('QG_1Layer_{}_{}.png'.format(N[0], count))
        if tstep % write_tstep[0] == 0:
            q = T.backward(q_hat, q)
            file.write(tstep, write_tstep[1])

    if __name__ == '__main__':

        file0 = HDF5File("QG_1Layer_{}.h5".format(N[0]), mode='w')
    
        par = {'plot_tstep': 100,
               'write_tstep': (100, {'q': [q]}),
               'file': file0}

        # Solve
        time_initial = time.time()
    
        integrator = RK4(T,  N=Flux_QG, update=update, **par)
        integrator.setup(dt)
        q_hat = integrator.solve(q, q_hat, dt, t_domain)
    
        time_final = time.time()

        generate_xdmf("QG_1Layer_{}.h5".format(N[0]))
        fftw.export_wisdom('QG1L.wisdom')

        # Compute energy and growth
        q      = T.backward(q_hat, q)
        #energy = comm.reduce(0.5*np.sum(U*U))

        print('Total time = ', (time_final - time_initial))
        
    `

**Comment by mikaem** [comment]
Thanks for the code. I had a good close look, and couldn't really find much. But I did manage to run it nicely in parallel doing two things:

1. `np.random.seed(rank)`, that you had commented out for some reason:-)
2. 

    if comm.Get_rank() == 1:
        generate_xdmf("QG_1Layer_{}.h5".format(N[0]))

which I realise was a mistake inherited from Ginzburg-Landau. With these two minor modifications it seems to run just fine, also in parallel:-)

**Comment by francispoulin** [comment]
Thank you Mikael for taking the time to look over my code.  I really appreciate it.

I am not sure why things are different but I agree, now that I run the code I get very reasonable solutoins, not like the strange one I posted yesterday.  Great news, thanks again!

I do understand why you inserted the if statement above, and that makes sense.  

Next, I'm going to work on changing the across channel direction to be polynomial, to allow for non-periodic flows, but I have one detail to ponit out.

When I specified the basis functions I did have domain(0, L[0]) and so on, but when I read the h5 file, it still reads the domain to be 0 to 2*pi.  Do I need to change the parameters when I write to a file?

**Comment by mikaem** [comment]
Hi Francis,
Just add `domain=T.domain()` when creating the `HDF5File` and I think that should do it.

**Comment by francispoulin** [comment]
Thanks for the quick reply.  I tried the following:

    `    file0 = HDF5File("QG_1Layer_{}.h5".format(N[0]), mode='w', domain=T.domain())`

but have an error that tensorproduct space has no attribute domain.  

Did I mess up the syntax?

**Comment by mikaem** [comment]
I was just guessing:-) The domain is for each Basis, not the tensorproductspace. So I'm now guessing you should provide two domains:-)

**Comment by francispoulin** [comment]
That was a good guess.

I found out that it's very easy to do it using the following syntax.

Cheers!

    `    domain = ((0., L[0]), (0., L[1]))
        file0 = HDF5File("QG_1Layer_{}.h5".format(N[0]), mode='w', domain=domain)
    `
